File: exercises/rb130/easy_2/ex2.rb
=begin

PROBLEM
input: 2 argument arrays
output: array of subarrays of matching element indices of each array
rules:
  explicit:
    - arrays have same number of elements
    - cannot use #zip

DATA
Arrays

ALGORITHM
- define a method `zip(array1, array2)`
- create an empty array `answer`
- iterate the size or the array times
  - push a 2 element array of the current iteration index from each array to `answer`
- return `answer`
=end

def zip(array1, array2)
  answer = []
  array1.size.times { |ind| answer << [array1[ind], array2[ind]] }
  answer
end

# def zip(array1, array2)
#   array1.each_with_index.with_object([]) do |(_, ind), answer|
#     answer << [array1[ind], array2[ind]]
#   end
# end

p zip([1, 2, 3], [4, 5, 6]) == [[1, 4], [2, 5], [3, 6]]

File: exercises/js210_small_problems/medium_1/ex2.js
/*

---PROBLEM---
input: number, integer n
output: number with last n digits rotated one digit to the left, moving first digit to the end

rules:
  - input a whole number
  - n a whole number
  - n <= total number of digits of number
  - rotate the last n digits one to the left
  - return value is a number

--DATA---
- array, iteration

---ALGORITHM---
- turn input numbr into an array of character
  - convert input numbr into a string
  - turn the string into an array of the characters
- iterate from length of array - n, n - 1 times
  - create a for loop from array's length - n, n - 1 times
- swap the current index with the current index plus 1
  - use [a, b] = [b - a] to mutate array while iterating through it
- rejoin the array into a number
  - join array back to string with empty string as delimieter
  - convert string to number & return

*/

function rotateRightmostDigits(number, n) {
  let array = String(number).split('');

  for (let i = array.length - n; i < array.length - 1; i += 1) {
    [array[i], array[i + 1]] = [array[i + 1], array[i]];
  }

  return Number(array.join(''));
}

console.log(rotateRightmostDigits(735291, 1));      // 735291
console.log(rotateRightmostDigits(735291, 2));      // 735219
console.log(rotateRightmostDigits(735291, 3));      // 735912
console.log(rotateRightmostDigits(735291, 4));      // 732915
console.log(rotateRightmostDigits(735291, 5));      // 752913
console.log(rotateRightmostDigits(735291, 6));      // 352917

